system/security/shared/seclib.hpp

```diff
@@ -409,64 +399,66 @@ enum secManagerType : int
+static const SecManagerFeature SMF_QueryDefaultPermission         = 0x0000004000000000;
+static const SecManagerFeature SMF_ClearPermissionsCache          = 0x0000008000000000;
+static const SecManagerFeature SMF_AuthenticateUser                 = 0x0000010000000000;
+static const SecManagerFeature SMF_QuerySecMgrType                  = 0x0000020000000000;
+static const SecManagerFeature SMF_QuerySecMgrTypeName              = 0x0000040000000000;
+static const SecManagerFeature SMF_LogoutUser                       = 0x0000080000000000;
+static const SecManagerFeature SMF_RetrieveUserData               = 0x0000100000000000;
+static const SecManagerFeature SMF_ALL_FEATURES                   = 0x00001FFFFFFFFFFF; // update to include all added feature bits
+using SecManagerFeatureSet = SecManagerFeature; // any number of features
+
 interface IEspSecureContext;
 interface ISecManager : extends IInterface
 {
-    virtual uint64_t queryFeatures(SecurityFeatureSupportLevel level, IEspSecureContext* secureContext = nullptr) const = 0;
-    inline bool checkFeatures(SecurityFeatureSupportLevel level, uint64_t mask, IEspSecureContext* secureContext = nullptr) const { return ((queryFeatures(level, secureContext) & mask) == mask); }
-    inline bool checkUnsafeFeatures(uint64_t mask, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Unsafe, mask, secureContext); }
-    inline bool checkSafeFeatures(uint64_t mask, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Safe, mask, secureContext); }
-    inline bool checkImplementedFeatures(uint64_t mask, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Implemented, mask, secureContext); }
+    // Returns the combination of SecManagerFeature values representing features with the indicated support level.
+    virtual SecManagerFeatureSet queryFeatures(SecurityFeatureSupportLevel level, IEspSecureContext* secureContext = nullptr) const = 0;
+    // Returns true if the given combinsation of SecManagerFeature values is a subset of the complete feature set for the indicated support level.
+    inline bool checkFeatures(SecurityFeatureSupportLevel level, SecManagerFeatureSet festureSet, IEspSecureContext* secureContext = nullptr) const { return ((queryFeatures(level, secureContext) & festureSet) == festureSet); }
+    inline bool checkUnsafeFeatures(SecManagerFeatureSet festureSet, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Unsafe, festureSet, secureContext); }
+    inline bool checkSafeFeatures(SecManagerFeatureSet festureSet, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Safe, festureSet, secureContext); }
+    inline bool checkImplementedFeatures(SecManagerFeatureSet festureSet, IEspSecureContext* secureContext = nullptr) const { return checkFeatures(SFSL_Implemented, festureSet, secureContext); }
```

Comment:
If you commoned up the types for SecManagerFeatureSet and SecUserFeatureSet could you put these functions in a common base interface?  ISecInterface or similar?

Reply:
i.e. if you just used uint64_t users would know it was just a common bitmask variable.

Reply:
It was uint64_t, and Russ asked for comments and name changes to clarify that the values are masks.

The types could be unified, but the user currently has no knowledge or use for the secure context. Is it worth complicating the user interface with unused parameters to eliminate four one line inline methods?

Reply:
how would using less typedefs and moving these up into a common base interface add unused parameters?

Reply:
oh I see what you mean about the unused parameter.  How is the context used in these get supported features functions?

Reply:
Moving these up may be a good excuse not to pass the secureContext if it's only being done in the name of consistency.

Reply:
When the manager wrapper (for future managers) is introduced, and the wrapper can reload the configuration, any manager call can trigger a reload. The context is being passed so that a potentially time consuming load can be tracked. I don't want people complaining about what should have been a trivial request (like returning a pre-calculated bit mask) taking too long without being able to record that the delay was caused by a manager re-configuration.

Reply:
It's also a consideration that a reload could actually load a different manager class, so we wouldn't be able to assume the bit masks are constants.

Reply:
You mean the security manager may reload itself because of a call to queryFeatures?  I could see tracking the time of a reload for the various calls that may cause a reload to happen.. although we're trying to move away from complexity.

But why would a reload be triggered by calling queryFeatures?  I think we should avoid such side effects.  And why would not passing a context imply that we were assuming static feature bits?